### DISCUSSION TITLE
Pass URL-based JSON as launch argument

### DISCUSSION THREAD

**Discussion by desentizised** [ChrisTitusTech/winutil#3045] [Ideas]
I wrote myself a batch script which enables me to launch winutil from cmd instead of having to launch a modern Windows shell explicitly and it gets its JSON instructions from a predefined URL. 

```
@echo off

:: Set variables for the URLs and paths
set "jsonUrl=https://example.com/custom.json"
set "jsonPath=%TEMP%\custom.json"

:: Check if the script is running as Administrator
net session >nul 2>&1
if %errorlevel% neq 0 (
    echo Requesting Administrator privileges...
    powershell -Command "Start-Process '%0' -Verb RunAs"
    exit /b
)

:: Download the JSON file using PowerShell
powershell -NoProfile -ExecutionPolicy Bypass -Command "Invoke-WebRequest -Uri '%jsonUrl%' -OutFile '%jsonPath%'"

:: Check if the JSON file was downloaded successfully
if exist "%jsonPath%" (
    :: Run the PowerShell script with the downloaded JSON file
    powershell -NoProfile -ExecutionPolicy Bypass -Command "iex ""& { $(irm christitus.com/win) } -Config '%jsonPath%'"""
    
    :: Clean up by deleting the downloaded JSON file
    del "%jsonPath%" 2>nul
) else (
    echo Failed to download the JSON configuration file.
)

exit /b
```

The main hurdle was finding the [documentation](https://christitustech.github.io/winutil/userguide/?h=config#config) for how to pass the `-Config`. At first I couldn't find anything because I was looking for keywords like "JSON" and "Import", then I looked at the main `.ps1` file and saw that it was indeed capable of ingesting a JSON on launch but that it was called "Config" in this context.

**Comment by MoRo816** [comment]
Sorry to bug months after you posted this.  My use case is probably simpler than yours.  I set up computers for home users a few times a week.  I have a NAS/Server with my tools and installers and WinUtil is one the first things I run.  I found the line at the bottom of the documentation you mentioned, but I can't seem to get the syntax right.

Currently my shortcut I run on a new client machine is set up with a 'Target' of: C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -ExecutionPolicy Bypass -Command "Start-Process powershell.exe -verb runas -ArgumentList 'irm https://christitus.com/win | iex'  

It runs fine and I can click the Import button and that will work fine, it's just a couple extra steps I'm trying to skip.  I think I'm putting a quotation mark in the wrong place or messing up the path in some basic way.

My config file is WinUtilConfig.json and it's in the same network folder where I have the shortcut saved.

Any help would be appreciated.

**Reply by desentizised** [reply]
Maybe @og-mrk's comment can help you out, I have yet to look into whether it can simplify my own approach above.

Our use-cases seem fairly similar. When I sit at a customer's PC and I know I need Chris' winutil I have a shortlink to a `.txt` which I know by heart (like url.ly/bat) so the webbrowser just comes up with the `.bat` text but it's not a `.bat` because downloading batch scripts off the internet isn't really feasible. Once I copy paste the text from the browser into a `.bat` everything's basically automatic and winutil starts with the checkboxes already set to my liking.

> My config file is WinUtilConfig.json and it's in the same network folder where I have the shortcut saved.

Basically in the line `powershell -NoProfile -ExecutionPolicy Bypass -Command "iex ""& { $(irm christitus.com/win) } -Config '%jsonPath%'"""` from my script you would need to replace `%jsonPath%` with your path, you don't need anything else from the script if the `.json` is already available locally or on a local network path. And if I understand @og-mrk correctly then this is now also possible with HTTP/HTTPS URLs.

**Comment by og-mrk** [comment]
You can now pass URL of your config file (whether the URL points to a remote server of yours, the Cloud, or any file hosting server), and WinUtil will resolve the URL and read the config without a problem. This was done in PR [#2815](https://github.com/ChrisTitusTech/winutil/pull/2815) by @MyDrift-user

**Reply by desentizised** [reply]
Very cool thanks!
